Add support for getting constellations and PrecessedGeocentric frame

eteq opened the pull request:
This PR does two related things to coordinates:

1. Adds a ``get_constellation`` convenience function and `SkyCoord` method. They do exactly what it sounds like: determine the constellation the coordinate is in and returns it.  
2. Adds a `PrecessedGeocentric` frame closely related to `GCRS`, but with added precession of the coordinate system.  This is important because it preserves the observer location as being relative to the Earth while allowing for precession.  This is critical for ``get_constellation`` because constellation boundaries are defined relative to the 1875 equinox, and the existing precess-able frames (FK4 and FK5) are barycentric, so you can't do things like asking what constellation the sun is in.  Adding this frame makes that possible and might be useful for other historical astronomy questions.

cc @adrn @astrofrog @taldcroft

taldcroft commented on astropy/coordinates/funcs.py:
```diff
@@ -176,8 +176,87 @@ def concatenate(coords):
+
+# global dictionary that caches repeatedly-needed info for get_constellation
+_constellation_data = {}
+def get_constellation(coord, short=False):
```
What about a more descriptive arg `short_name`?

eteq replied:
Good idea - will change.

taldcroft commented on astropy/coordinates/funcs.py:
```diff
@@ -176,8 +176,87 @@ def concatenate(coords):
+_constellation_data = {}
+def get_constellation(coord, short=False):
+    """
+    Determines the constellation(s) of the coordinates this `SkyCoord`
```
The coord can also be a low-level coord frame object, not just `SkyCoord`.

eteq replied:
Oops, yeah.  The need for this change (and a similar change in "Returns") also means the bit I suggested on astropy-dev probably isn't worth it for this... Will fix

taldcroft commented on astropy/coordinates/tests/test_funcs.py:
```diff
@@ -49,3 +51,20 @@ def test_concatenate():
+    from .. import ICRS, FK5
+    from ..funcs import get_constellation
+
+    inuma = ICRS(9*u.hour, 65*u.deg)
```
Might be worth including a specific test for Boötes, it being the unicode oddball.

eteq replied:
Ah, good point.  Added in version I'll push up shortly.

taldcroft commented on astropy/coordinates/builtin_frames/gcrs.py:
```diff
@@ -69,5 +68,59 @@ class GCRS(BaseCoordinateFrame):
+# the current implementation it goes through ICRS (like CIRS)
+
+
+class PrecessedGeocentric(BaseCoordinateFrame):
```
I don't see any specific tests for this new frame (besides use in constellation code).

eteq replied:
that's right - I left it as implicit via the constellation code.  But I suppose you're right an explicit one is a good idea, too.

taldcroft commented:
@eteq - finished review.

I never would have guessed that the formal definition of constellation boundaries is rectilinear like this, but I googled it and now I believe.

astrofrog commented:
@eteq - I think it would be good to add an option to specify which set of constellations to return (defaulting to Roman). Maybe we will want to integrate ones from other cultures one day?

eteq commented:
Ok, I think this is all set - I've addressed @taldcroft's comments, and added an option for future other constellations as @astrofrog suggested (note that the set here is not "Roman", but rather the IAU's official "modern" constellations - it differs some from Greco-Roman constellations in that there are a bunch more in the south where the classical era didn't have names for the constellations).

As you say, @taldcroft, I was also surprised to find this is how they're laid out... But I suppose it makes sense for 1930 there were no computers to try to figure out how curvy boundaries map to the point you're looking at, so square segments are just easier to look up.
